proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,521 @@
+
+This limitation in expressivity is emphasized by the fact that property wrappers were originally sought out to abstract away such patterns.  As a result, elegant APIs are undermined by this limitation. Not only is this limiting users by forcing them to carefully read documentation, which may not cover a specific use case, to make sure no invariants have been violated, but it also limits API authors in what they can create. That is, API authors can't use property-wrapper types in closure parameters nor can code be separated into functions that accept property wrapper syntax:
+
+```swift
+extension Percentage {
+
+  func modify(
+    inSeconds seconds: Int,
+    block: @escaping (inout Clamped<Int>) -> Void
+  ) { ... }
+  
+}
+
+
+let myPercentage = Percentage(percent: 50)
+
+myPercentage
+  .modify(inSeconds: 3) { percent in
+    percent.wrappedValue = 100
+    //    ^~~~~~~~~~~~ 
+    // Again, we have to access
+    // count through 'wrappedValue'.
+  }
+```
+
+In fact, establishing custom behavior on closure parameters is really powerful. For example, if such a feature were supported, it could be used in conjunction with [Result Builders](https://github.com/apple/swift-evolution/blob/main/proposals/0289-result-builders.md) to expose data managed by a 'component' type. For instance, in SwiftUI [`ForEach`](https://developer.apple.com/documentation/swiftui/foreach) could leverage this feature to expose the mutable state of its data source to its 'content' closure. This would enable users to more easily work with the data source itself inside the closure instead of accessing the original property, which is particularly painful when working with collections, as shown in this example:
+
+```swift
+struct MyView : View {
+
+  // A simple Shopping Item that includes
+  // a 'quantity' and a 'name' property.
+  @State 
+  private var shoppingItems: [Item]
+
+
+  var body: some View {
+    ForEach(0 ..< shoppingItems.count) { index in
+  
+      TextField("Enter the item's name...", $shoppingItems[index].name)
+      
+    }
+  }
+  
+}
+```
+
+we would — with an appropriate initializer — be able to simplify the above code, reducing boilerplate as a result:
+
+```swift
+struct MyView: View {
+
+  @State 
+  private var shoppingItems: [Item]
+
+
+  var body: some View {
+    ForEach($shoppingItems) { (@Binding shoppingItem) in
+    
+      TextField("Enter the item's name...", $shoppingItem.name)
+      
+    }
+  }
+  
+}
+```
+
+
+## Proposed solution
+
+We propose to extend the contexts where application of property-wrapper types is permitted. Namely, application of such types will be allowed on function and closure parameters:
```

Comment:
Is there any reason why we removed the example that was here or was it removed by accident?

Reply:
I removed it because it was really similar to the example above with `(@Binding shoppingItem)` but it was less compelling, so I didn't think it added anything. One thing we could do instead is move the `Binding` example here and explain it a little more, and just leave the "before" examples in the motivation

Reply:
Oh, okey. I saw the colon right after "parameters" and I thought that I might have deleted that by accident. As far as the moving of the "after" examples to the Proposed Solution section, I think that's a really good idea. There's, also, not a lot of space between the "before" examples and the Proposed Solution section, so I think it's going to work out really well!